### PR TITLE
PPS detector mapping records separation and serialization

### PR DESCRIPTION
**Pull request by brzozia**
#### PR description:

This PR is a first step in a way to resolve an issue from https://github.com/cms-sw/cmssw/pull/40765 - Transfer of all PPS detectors' mappings from XML to CondDB.

In this PR:
* Introduction of new record: [TotemAnalysisMaskRcd.cc](https://github.com/CTPPS/cmssw/pull/79/files#diff-8e15d43975e75857b41d01b6bfd950e6e7a2a87b203ea58acce2cad3100e8490) - separated from TotemReadoutRcd;
* Serialization of classes which objects are intended to be placed in the CondDB (and inner classes of those) - [TotemAnalysisMask.h](https://github.com/CTPPS/cmssw/pull/79/files#diff-0266d70e3769fc2f3d773fb4a84ae048b23986dcff928e5a65d104dfd87770b9), [TotemDAQMapping.h](https://github.com/CTPPS/cmssw/pull/79/files#diff-c70098b419b2deac1b51936061ac94c788496fa4724139ff74c276006e161b56), [TotemFramePosition.h](https://github.com/CTPPS/cmssw/pull/79/files#diff-a64176b2f94eec46393738ac968202e6055cc1795695836c6f5f28c62dd63e7b), [TotemSymbId.h](https://github.com/CTPPS/cmssw/pull/79/files#diff-d3e99bdc9a0a36bab8559221eb41b4797245dd34799c91c6aced88e3ce6853d4));
* Writer plugin ([WriteCTPPSTotemDAQMappingMask.cc](https://github.com/CTPPS/cmssw/pull/79/files#diff-140c17cd5e469a3e3692f7f3e397d10512b8b9e95d1038ac986f3737902f8475)) and config ([write-ctpps-totem_daqmap_cfg.py](https://github.com/CTPPS/cmssw/pull/79/files#diff-e836bfb51a0ad8796ba77e01102e7487caaedfa09c9d56a28f7b1c59d4b8b5d0)) for reading CTPPS XML files with mappings and placing objects made from those in SQLite;
* Updates of printer plugin ([PrintTotemDAQMapping.cc](https://github.com/CTPPS/cmssw/pull/79/files#diff-2f22f8e72a591e6be95ba3f5388a9a837648b9153fb5ba8e7bfec2b1cc2211ff)) which prints records (TotemReadoutRcd, TotemAnalysisMaskRcd) content;

#### PR validation:

Validation has been made with use of new plugin ([CalibPPS/ESProducers/plugins/WriteTotemDAQMapping.cc](https://github.com/CTPPS/cmssw/pull/79/files#diff-276299009e7ad4e9336945c66d7cfcb5b6ec85d7835707bcd3096bcfb382c7dc)) which reads data from SQLite/XML file and drops them to text file.
All data from XML was dropped to SQLite in separate records and with different IOVs.
```
python3 CondTools/CTPPS/test/script-ctpps-write-many-XML-to-SQLite.py
```
Then (using new plugin) objects were read from SQLite and dropped to text file.
```
python3 CalibPPS/ESProducers/test/script_test_many_writeTotemDAQMapping.py
```
Also data from XML files was read and dropped to other text file (change flag inside script and then run).
```
python3 CalibPPS/ESProducers/test/script_test_many_writeTotemDAQMapping.py
```
Then files were compared and their content was the same, as expected eg.:
``` 
diff all_diamonds_db.txt all_diamonds_xml.txt
```

### CODE REVIEW

File: CondFormats/PPSObjects/src/classes.h [hunk 1]
```diff
@@ -20,5 +20,18 @@ namespace CondFormats_CTPPSPixelObjects {
     LHCInterpolatedOpticalFunctionsSet lhc_iofs;
     LHCInterpolatedOpticalFunctionsSetCollection lhc_iofsc;
     PPSPixelTopology pps_pt;
+
+    
   };
-}  // namespace CondFormats_CTPPSPixelObjects
+}
+
+// namespace {
```
Comment (grzanka) [inline CondFormats/PPSObjects/src/classes.h:28]: remove if not needed

**Comment by grzanka** [comment]
In general looks good, ask @JanChyczynski for a review ! Squash commits may be needed.
Have you tried `runTheMatrix` tests here ?

**Comment by JanChyczynski** [comment]
Please run
```
scram build code-checks
scram build code-format
```
And apply the changes

File: CondTools/CTPPS/test/write-ctpps-totem_daqmap_cfg.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    lastValue = cms.uint64(min_iov),
+    interval = cms.uint64(1)
+)
+
+process.MessageLogger = cms.Service("MessageLogger",
+  destinations = cms.untracked.vstring('cout'),
+  cout = cms.untracked.PSet(
+      threshold = cms.untracked.string('INFO'),
+      
+  ))
```
Comment (JanChyczynski) [inline CondTools/CTPPS/test/write-ctpps-totem_daqmap_cfg.py:22]:
```suggestion
      )
  )
```

File: CalibPPS/ESProducers/plugins/PrintTotemDAQMapping.cc [hunk 3]
```diff
@@ -30,33 +31,45 @@ class PrintTotemDAQMapping : public edm::one::EDAnalyzer<> {
...
   // get mapping
-  auto const &mapping = es.getData(mappingToken_);
+  if (auto mappingHandle = es.getHandle(mappingToken)) {
+    auto const &mapping = *mappingHandle;
+    // print mapping
+    for (const auto &p : mapping.VFATMapping)
+      edm::LogInfo("PrintTotemDAQMapping mapping") << "    " << p.first << " -> " << p.second;
+
+    for (const auto &p : mapping.totemTimingChannelMap)
+      edm::LogInfo("PrintTotemDAQMapping channel mapping") << "    " << p.first << " plane " << p.second.plane << " channel " << p.second.channel;
```
Comment (JanChyczynski) [inline CalibPPS/ESProducers/plugins/PrintTotemDAQMapping.cc:55]: After addressing the comment in `WriteCTPPSTotemDAQMappingMask.cc` about printing you could use the overloaded `<<` operators here.

Some more places like this in this file

File: CalibPPS/ESProducers/test/scipt_test_many_writeTotemDAQMapping.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+            if from_db:
+                content = re.sub(r'process.es_prefer_totemTimingMapping =.*', 
+                                f'process.es_prefer_totemTimingMapping = cms.ESPrefer("PoolDBESSource", "", TotemReadoutRcd=cms.vstring(f"TotemDAQMapping/{file_content["subSystemName"]}"))', content)
+            else:                
+                content = re.sub(r'process.es_prefer_totemTimingMapping =.*', 
+                                f'process.es_prefer_totemTimingMapping = cms.ESPrefer("TotemDAQMappingESSourceXML","totemDAQMappingESSourceXML",TotemReadoutRcd=cms.vstring("TotemDAQMapping/{file_content["subSystemName"]}"))', content)            
+          
+            f.seek(0)
+            f.write(content)
+            f.truncate()
```
Comment (JanChyczynski) [inline CalibPPS/ESProducers/test/scipt_test_many_writeTotemDAQMapping.py:155]: Same issues as in [script-ctpps-write-many-XML-to-SQLite.py](https://github.com/CTPPS/cmssw/pull/79/files#diff-0e4fc4d568e2cae19bf49dd0c7cad074f6bc12fb0e8ed3cf69511660cd89711b):
* too long lines
* hard to understand: split to subfunctions or comment

**Review by JanChyczynski — CHANGES_REQUESTED** [review]
Apart from some mostly minor things looks good

**Comment by grzanka** [comment]
@brzozia please squash the commits. I think we are ready to go with an official PR, assuming that runTheMatrix shows no problems

**Review by JanChyczynski — APPROVED** [review]
looks good

**Pull request merged**